File: src/tools/tester.ts
import { chromium, Browser, Page } from "playwright";
import { Workspace } from "../workspace.js";

export class PlaywrightTester {
  constructor(private workspace: Workspace) {}

  async runTest(url: string, screenshotPath: string) {
    const resolvedScreenshot = this.workspace.resolve(screenshotPath, false);
    const browser = await chromium.launch();
    const page = await browser.newPage();
    await page.goto(url);
    await page.screenshot({ path: resolvedScreenshot });
    await browser.close();
    return resolvedScreenshot;
  }

  async executePlaywrightCommand(command: string) {
    // This could be used to run 'npx playwright test'
    // For now, we'll assume the SandboxExecutor can handle the command line part,
    // and this class provides high-level browser interactions if needed.
  }
}

File: tester/tester.go
package tester

import (
	"fmt"
	"os/exec"

	"github.com/donomii/j/types"
	"github.com/donomii/j/workspace"
)

type PlaywrightTester struct {
	workspaceRoot types.WorkspaceRoot
}

func New(workspaceRoot types.WorkspaceRoot) *PlaywrightTester {
	return &PlaywrightTester{workspaceRoot: workspaceRoot}
}

func (t *PlaywrightTester) RunTest(targetURL, screenshotPath string) (string, error) {
	resolvedScreenshot, err := workspace.Resolve(t.workspaceRoot, screenshotPath, false)
	if err != nil {
		return "", fmt.Errorf("prepare browser screenshot path %q: %w", screenshotPath, err)
	}
	cmd := exec.Command("npx", "playwright", "screenshot", "--browser=chromium", targetURL, resolvedScreenshot)
	out, err := cmd.CombinedOutput()
	if err != nil {
		return "", fmt.Errorf("capture browser screenshot from %q into %q: %w: %s", targetURL, resolvedScreenshot, err, out)
	}
	return resolvedScreenshot, nil
}
